enclave/src/main.rs

```diff
@@ -0,0 +1,180 @@
+
+    Ok(())
+}
+
+/// Performs the following BPF-related operations:
+/// - loading BPF programs
+/// - pinning BPF maps in BPFFS
+/// - pinning BPF programs in BPFFS
+/// - attaching BPF programs, creating links
+/// - pinning links in BPFFS
+///
+/// All entities pinned in BPFFS have the dedicated directory signed with a
+/// timestamp. The reason behind it is to be able to still keep running
+/// previous instances of BPF programs while we are in the process of loading
+/// new programs. This is done to ensure that **some** instance of BPF programs
+/// is always running and that containers are secured.
+///
+/// TODO: The concept described above still has one hole - the contents of old
+/// BPF maps is not migrated in any way. We need to come up with some sane copy
+/// mechanism.
+fn load_programs(path_base_ts: path::PathBuf) -> Result<(), LoadProgramError> {
+    let skel_builder = EnclaveSkelBuilder::default();
+    let open_skel = skel_builder.open()?;
+    let mut skel = open_skel.load()?;
+
+    let mut path_map_runtimes = path_base_ts.join("map_runtimes");
+    skel.maps_mut().runtimes().pin(&mut path_map_runtimes)?;
+
+    init_runtimes(skel.maps_mut().runtimes())?;
+    
+    let path_map_containers = path_base_ts.join("map_containers");
+    skel.maps_mut().containers().pin(path_map_containers)?;
+
+    let path_program_fork = path_base_ts.join("prog_fork");
+    skel.progs_mut().sched_process_fork().pin(path_program_fork)?;
+
+    let path_program_clone = path_base_ts.join("prog_clone_audit");
+    skel.progs_mut().clone_audit().pin(path_program_clone)?;
+
+    let path_program_syslog = path_base_ts.join("prog_syslog_audit");
+    skel.progs_mut().syslog_audit().pin(path_program_syslog)?;
+
+    let mut link_fork = skel.progs_mut().sched_process_fork().attach()?;
+    let path_link_fork = path_base_ts.join("link_fork");
+    link_fork.pin(path_link_fork)?;
+
+    let mut link_clone = skel.progs_mut().clone_audit().attach_lsm()?;
+    let path_link_clone = path_base_ts.join("link_clone_audit");
+    link_clone.pin(path_link_clone)?;
+
+    let mut link_syslog = skel.progs_mut().syslog_audit().attach_lsm()?;
+    let path_link_syslog = path_base_ts.join("link_syslog_audit");
+    link_syslog.pin(path_link_syslog)?;
+
+    Ok(())
+}
+
+#[derive(thiserror::Error, Debug)]
+enum CleanupError {
+    #[error("regex compilation error")]
+    RegexError(#[from] regex::Error),
+
+    #[error("I/O error")]
+    IOError(#[from] std::io::Error),
+
+    #[error("could not convert path to string")]
+    PathToStrConvError,
+}
+
+/// Removes all old BPF entities (programs, maps, links) from BPFFS, to stop
+/// the execution of old BPF programs. All directories with timestamp lower
+/// than the current one get removed.
+fn cleanup(path_base: path::PathBuf, dirname: &String) -> Result<(), CleanupError> {
+    let rx = regex::Regex::new(format!(r#"{}"#, dirname).as_str())?;
+    let paths = fs::read_dir(path_base)?;
+
+    for entry in paths {
```

Comment:
I cannot recall for sure, but you may be able to simplify this to:
```rust
for entry in fs::read_dir(path_base)? {
   //--snip--
}
```
There might be a borrow-checker restriction, but I would check.

Reply:
It works